Warnings collection.

hurufu opened the pull request:
After recent discussion #2599, I've decided to start collecting useful warnings that can be applied during compilation. If you have good ideas for warnings, please post them here.

Relevant issues: #2600 #2601.

If you want to test, just import `warnings` module:

```prolog
:- use_module(library(warnings)).
```

And that's all.

triska commented on src/lib/warnings.pl:
```diff
@@ -0,0 +1,37 @@
+    prolog_load_context(term_position, position_and_lines_read(_,L)),
+    append(["% Warning: ", Format, " at line ~d of ~a~n"], FullFormat),
+    append(Vars, [L,F], AllVars),
+    format(S, FullFormat, AllVars),
```
Is this:

<pre>
phrase_to_stream(("% Warning: ",format_(Format, Vars),format_(" at line ~d of ~a~n", [L,F])), O)
</pre>

?

hurufu replied:
Ah, indeed, `phrase` will be much more elegant.

hurufu replied:
Fixed

hurufu commented:
Tests are failing, because of duplicated warning message, I'll try to come up with minimal example.

hurufu commented:
Currently singleton variables warning is done in loader.pl during goal expansion, if this PR will be accepted, I can move that code to `library(warnings)`.

triska commented:
Conceptually very interesting! This is an aspect that I think would also generally benefit from more thought during development: Which parts of the system must be provided as foundation, and which can be built on top?

hurufu commented:
Ideas for warnings:

- [x] Warn if user tries to re-define built-in static predicate
- [ ] Warn if user-provided `warnings:output_stream/1` misbehaves.
- [ ] Warn if a non-list is found eg.: `[a,b,c|d]`.
- [ ] Detect used, but not visible predicates + suggest which module to load (needs some [additional support](https://github.com/mthom/scryer-prolog/issues/2601#issuecomment-2395014262) from loader)
- [ ] Warn about duplicated cuts (`!/0`) in a single predicate body

This list will be expanded in the future. Regarding implementation... if community thinks that this is a good idea, I will implement them. Currently I only gather ideas.

triska commented:
One question that may be worth taking into account: What makes most programs wrong? I think the type tests `integer/1`, `atom/1` are candidates for this: Their use makes many programs yield wrong&nbsp;answers. The safe type tests `integer_si/1` etc. from `library(si)` could be automatically suggested in such cases.

triska commented on src/lib/warnings.pl:
```diff
@@ -1,39 +1,50 @@
+    nonvar(G),
+    unsound_type_test(G),
+    functor(G, F, 1),
+    warn_fail("~q is a constant source of bugs, use ~a_si/1 from library(si)", [F/1,F]).
```
maybe "of wrong results"?

triska commented:
There seems to be a pattern in the expansion code, so maybe this code can be generated from shorter declarative descriptions of warnings?

triska commented:
#2189 may be a candidate to address with this technique.

hurufu commented:
I've added quite interesting and less trivial warning. It now detect missing meta_predicate declaration. I don't like how I implemented it, main that hook doesn't seem elegant to me. There is a lot of useful code in loader, but it is hard to use it in different contexts.

hurufu commented:
I've copied code from #2458, that particular task should be fixed in the compiler itself, but Prolog code is useful enough to give it second life here.
